src/Stratis.Features.FederatedPeg/TargetChain/MempoolCleaner.cs

```diff
@@ -0,0 +1,131 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.AsyncWork;
+using Stratis.Bitcoin.Features.MemoryPool;
+using Stratis.Bitcoin.Utilities;
+using Stratis.Features.FederatedPeg.Interfaces;
+using Stratis.Features.FederatedPeg.Wallet;
+
+namespace Stratis.Features.FederatedPeg.TargetChain
+{
+    /// <summary>
+    /// This component is responsible clearing completed withdrawals from the mempool.
+    /// </summary>
+    public interface IMempoolCleaner
+    {
+        /// <summary>
+        /// Starts the cleaning of the mempool every N seconds.
+        /// </summary>
+        void Start();
+
+        /// <summary>
+        /// Stops the cleaning of the mempool.
+        /// </summary>
+        void Stop();
+    }
+
+    public class MempoolCleaner : IMempoolCleaner, IDisposable
+    {
+        /// <summary>
+        /// How often to trigger the cleaning of the mempool.
+        /// </summary>
+        private static readonly TimeSpan TimeBetweenQueries = TimeSpans.Minute;
+        private readonly ILogger logger;
+        private readonly MempoolOrphans mempoolOrphans;
+        private readonly ICrossChainTransferStore store;
+        private readonly IAsyncProvider asyncProvider;
+        private readonly INodeLifetime nodeLifetime;
+        private readonly IFederationWalletManager federationWalletManager;
+
+        private IAsyncLoop asyncLoop;
+
+        public MempoolCleaner(
+            ILoggerFactory loggerFactory,
+            MempoolOrphans mempoolOrphans,
+            ICrossChainTransferStore crossChainTransferStore,
+            IAsyncProvider asyncProvider,
+            INodeLifetime nodeLifetime,
+            IFederationWalletManager federationWalletManager)
+        {
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+            this.mempoolOrphans = Guard.NotNull(mempoolOrphans, nameof(mempoolOrphans));
+            this.store = Guard.NotNull(crossChainTransferStore, nameof(crossChainTransferStore));
+            this.asyncProvider = Guard.NotNull(asyncProvider, nameof(asyncProvider));
+            this.nodeLifetime = Guard.NotNull(nodeLifetime, nameof(nodeLifetime));
+            this.federationWalletManager = Guard.NotNull(federationWalletManager, nameof(federationWalletManager));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        private IEnumerable<Transaction> CompletedTransactions(IEnumerable<Transaction> transactionsToCheck)
```

Comment:
I don't think this quite covers the cases. For consolidating transactions for example:

- 2 nodes both fully sign a tx at the same time. Both txs are broadcasted. 1 becomes an orphan.

So in this case the orphaned transaction has a different hash to the completed transactions.

It's hard though because the transaction could be lingering in the FWM

Reply:
These lingering transactions can probably be identified by the fact that one or more of their inputs will eventually be in-use by confirmed transactions?

Reply:
Done.